<a href="https://colab.research.google.com/github/Ryan-M-Smith/CS315/blob/main/HW09/hw09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# HW09
Ryan Smith

In [2]:
import numpy as np
import math

## 1. Caesar Square

The string of letters contains 36 characters. Because 36 is a perfect square, we can arrange these characters into a $6\times6$ matrix - a Caesar Square. When we place the characters in the matrix and transpose it, we get:

In [3]:
input_str = "NARRFSECCNTCXAIOUITEPSSIISHHEOSAEIAK"
letters = [l for l in input_str]
n = int(math.sqrt(len(letters)))

# Reshape the letters into an n by n matrix
matrix = np.array(letters).reshape(n, n)
matrix.transpose()
print(matrix)

[['N' 'A' 'R' 'R' 'F' 'S']
 ['E' 'C' 'C' 'N' 'T' 'C']
 ['X' 'A' 'I' 'O' 'U' 'I']
 ['T' 'E' 'P' 'S' 'S' 'I']
 ['I' 'S' 'H' 'H' 'E' 'O']
 ['S' 'A' 'E' 'I' 'A' 'K']]


Message: “NEXT IS A CAESAR CIPHER NO SHIFT USE ASCII OK”

## 2. Caesar Cipher (no shift)

Using the hint from problem 1, we know that this series of integers represents a Caesar Cipher with no shift. To decode it, we can simply convert each number from its ASCII value to the corresponding character. This gives us:

In [5]:
ciphertext = [
    84, 104, 101, 32, 110, 101, 120, 116, 32, 99, 108, 117, 101, 32, 119, 97, 115, 32, 99, 114,
    101, 97, 116, 101, 100, 32, 119, 105, 116, 104, 32, 97, 32, 50, 32, 98, 121, 32, 50, 32, 109,
    97, 116, 114, 105, 120, 32, 82, 111, 119, 32, 111, 110, 101, 32, 105, 115, 32, 52, 32, 97,
    110, 100, 32, 50, 32, 82, 111, 119, 32, 116, 119, 111, 32, 105, 115, 32, 51, 32, 97, 110, 100,
    32, 49
]
plaintext = list(map(chr, ciphertext))
print("".join(plaintext))

The next clue was created with a 2 by 2 matrix Row one is 4 and 2 Row two is 3 and 1


Message: The next clue was created with a 2 by 2 matrix. Row one is 4 and 2. Row two is 3 and 1.

## 3. Matrix Encryption

Using the hint from problem 2, we know that this is a matrix encryption problem. The hint gives us our $A$ matrix:

$
A = \left[\begin{matrix}
4 & 2 \\ 3 & 1
\end{matrix}\right]
$

Our $C$ matrix will then be constructed from the numbers provided.

We can solve this by:

$
AB = C \implies B = A^{-1}C
$

With $B$ calculated, we can then decode the resulting Caesar Cipher like we did in problem 2.

In [12]:
ciphertext = [
    556, 480, 568, 294, 562, 452, 684, 698, 324, 652, 614, 658, 524, 602, 614, 646,
    192, 698, 592, 524, 330, 518, 554, 500, 490, 552, 468, 620, 250, 582, 516, 572,
    522, 238, 480, 362, 344, 385, 179, 389, 323, 457, 465, 194, 435, 408, 444, 377,
    398, 410, 424, 128, 468, 393, 377, 197, 358, 391, 351, 342, 392, 335, 410, 157,
    410, 363, 402, 365, 151, 337
]

# Calculate B
A = np.array([[4, 2], [3, 1]])
A_inv = np.linalg.inv(A)
C = np.array(ciphertext).reshape(2, -1)
B = A_inv.dot(C).astype(int)

# Decode the message
B = B.flatten()
plaintext = list(map(chr, B))
print("".join(plaintext))

The last message was created with an RSA public key of e=103, n=50657.


Message: The last message was created with an RSA public key of $e=103$, $n=50657$.

## 4. RSA Encryption

Using the hint from problem 3, we know the last message was encrypted using RSA, and we are given $e=103$ and $n=50657$. In order to decrypt RSA, we need to compute $S(C) = C^d \ \mathrm{mod} \ n$, where $C$ is the ciphertext. In this case, the ciphertext is provided as a list of integers, so we'll apply this formula to each character.

However, we have been given $e$ and $n$, but not $d$. To find $d$, we must first find $\phi = (p - 1)(q - 1)$. From there, we can solve for $d$ such that:

$
ed \equiv 1 \ \mathrm{mod} \ \phi
$

Thus, $d$ is the _mod inverse_ of $e$ and $\phi$.

Because this is a small example, we can "crack" the encryption by figuring out $p$ and $q$ - the prime numbers whose product is $n$. Using [a prime factorization calculator](https://www.calculatorsoup.com/calculators/math/prime-factors.php), we can see that the prime factorization is $179 \times 283$. Thus:

$
p = 179 \\
q = 283
$

From here, we can perform the decyption:

In [13]:
def mod_inverse(e: int, phi: int) -> int:
  e %= phi

  for x in range(1, phi):
    if (e * x) % phi == 1:
      return x

  return 1

In [14]:
e = 103
n = 50657
p = 179
q = 283

# Calculate d
phi = (p - 1) * (q - 1)
d = mod_inverse(e, phi)

# Decrypt
ciphertext = [
    10786, 3034, 20400, 12827, 11787, 22288, 45253, 15845, 27221, 22288, 45253,
    40195, 3034, 20400, 39050, 25123, 22852, 15801, 3034, 15845, 22852, 28772,
    22288, 4804, 27259, 22852, 38364, 3034, 16811, 42019, 27221, 27259, 45253,
    27259, 1721, 22852, 10786, 44541, 22852, 44954, 32100, 19877, 22852, 7661,
    40817, 39163, 16180, 34613, 22852, 6459, 3034, 11787, 22852, 41659, 15845,
    27221, 27221, 611, 38364, 11787, 27259, 1721, 40195, 45253, 45324, 22852,
    27259, 20400, 45253, 27259, 11787, 22852, 45253, 28772, 27259, 22852, 41659,
    3034, 27221, 27221, 3034, 897, 40195, 20400, 12827, 22852, 16811, 27259,
    39050, 39050, 22288, 12827, 27259, 22852, 27259, 38496, 22288, 38364, 45253,
    27221, 11222, 22852, 40195, 20400, 45253, 3034, 22852, 3270, 3034, 3034,
    1721, 27221, 27259, 49101, 22852, 3855, 30551, 11787, 3034, 41659, 22852,
    584, 22852, 42019, 27221, 27259, 22288, 39050, 27259, 22852, 22288, 39050,
    39050, 40195, 12827, 20400, 22852, 22288, 27221, 27221, 22852, 45253, 28772,
    27259, 22852, 42019, 3034, 40195, 20400, 45253, 39050, 45324, 22852, 37276,
    22852, 41659, 40195, 20400, 40195, 39050, 28772, 27259, 1721, 22852, 7661,
    40817, 39163, 16180, 22852, 4804, 27259, 11787, 39050, 40195, 3034, 20400,
    22852, 38364, 3855
]
plaintext: list[chr] = []

for num in ciphertext:
  plaintext.append(chr((int(num) ** d) % n))

word = "".join(plaintext)
print(word)

Congratulations! You have completed CS 315 HW09. For full-credit, enter the following message exactly into Moodle: "Prof K please assign all the points, I finished HW09 version c"


Message: Congratulations! You have completed CS 315 HW09. For full-credit, enter the following message exactly into Moodle: "Prof K please assign all the points, I finished HW09 version c"